### ISSUE TITLE
IllegalArgumentException (shape issue) when calling datasetIterator.next()

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/deeplearning4j#1254]
Experiencing an IllegalArgumentException when attempting to train a convolutional net. After successfully narrowing any possible problems with the dataset, I noticed that it's happening when I call `next()` on my dataset iterator.

Relevant code (gist below):
```
while(dataSetIterator.hasNext) {
            val next: DataSet = dataSetIterator.next()
            next.scale()
            val trainTest = next.splitTestAndTrain(splitTrainNum, new Random(seed))  // train set that is the result
            val trainInput = trainTest.getTrain  // get feature matrix and labels for training
            testInputBuilder += trainTest.getTest.getFeatureMatrix
            testLabelsBuilder += trainTest.getTest.getLabels
            model.fit(trainInput)
        }
```

Relevant error:
```
Exception in thread "main" java.lang.IllegalArgumentException: Index [1] must not be >= shape[d].
	at org.nd4j.linalg.api.shape.Shape.getOffset(Shape.java:484)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.putScalar(BaseNDArray.java:1071)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.putScalar(BaseNDArray.java:1050)
	at org.deeplearning4j.datasets.canova.RecordReaderDataSetIterator.getDataSet(RecordReaderDataSetIterator.java:210)
	at org.deeplearning4j.datasets.canova.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:150)
	at org.deeplearning4j.datasets.canova.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:288)
	at org.deeplearning4j.datasets.canova.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:42)
	at org.deeplearning4j.examples.convolution.BernieFaces$.main(BernieFaces.scala:151)
	at org.deeplearning4j.examples.convolution.BernieFaces.main(BernieFaces.scala)
```

I've included a Gist that reproduces the problem: https://gist.github.com/crockpotveggies/e1d1607c2f70529181ee

On the surface this doesn't appear related to usage with Scala. It could very well be a shaping issue, but after playing with the parameters and constructors, I'm not so sure. The gist contains details about each image. There are 9530 images total, with 8 labels and each image is 100x100.

If you need a dataset to test this network, I'm more than happy to PM a link to a private repo for reproducibility.

**Comment by AldoMarzullo** [comment]
Hi, I am dealing with the same issue. This is my code if it can help
http://pastebin.com/Dmga0sfU

**Comment by AlexDBlack** [comment]
@crockpotveggies looks like your issue is simply that you are passing false as the argument to appendLabels:
val recordReader = new ImageRecordReader(numRows, numColumns, nChannels, false, labels)
but then you are saying that the label is present at index 10000:
val dataSetIterator: DataSetIterator = new RecordReaderDataSetIterator(recordReader, 10000, 8)
try using appendLabels = true

**Comment by AlexDBlack** [comment]
@AldoMarzullo your issue appears to be due to a different cause. I've been unable to reproduce this locally... can you give me some images to run your code with? Talk to me on gitter if you don't want to share them publicly.

**Comment by crockpotveggies** [comment]
Ah yes I did try that actually. That suggestion originated from our Gitter chat. I found two problems when setting `appendLabels = true`:

1) If I pre-pass the labels then I actually end up with 16 labels in scope (even thought there are only 8). I think it might be an API misunderstanding, but in the Java package it says to pass an ArrayList of "indexed" labels. So what I've done is set the directory names to `0-7` and define their labels in the ArrayList. 
2) More importantly, if I either pre-pass labels or set them to true, I get this error:

```
Exception in thread "main" java.lang.IllegalArgumentException: Unable to split on size larger than the number of rows
	at org.nd4j.linalg.dataset.DataSet.splitTestAndTrain(DataSet.java:621)
	at org.deeplearning4j.examples.convolution.BernieFaces$.main(BernieFaces.scala:150)
	at org.deeplearning4j.examples.convolution.BernieFaces.main(BernieFaces.scala)
```

**Comment by crockpotveggies** [comment]
One more thing: if I reduce the size of `splitTrainNum` being passed into `splitTestAndTrain` to a ridiculously small number, it has no effect on that same `IllegalArgumentException`. It might to help to change `rows` to say `number of examples` - I'm more than happy to submit a PR for this. However, I'm still stumped as to why it would fail. @AlexDBlack

**Comment by AldoMarzullo** [comment]
@AlexDBlack I am using the public available dataset [DRIVE](http://www.isi.uu.nl/Research/Databases/DRIVE/) and [RIMONE](http://medimrg.webs.ull.es/research/retinal-imaging/rim-one/), I attached two sample images. 

[drive_sample.zip](https://github.com/deeplearning4j/deeplearning4j/files/170494/drive_sample.zip)
I also tried with two random images downloaded from the web.

**Comment by nyghtowl** [comment]
@crockpotveggies A note, batchSize is typically between 32 and 128. Too large and you may not be updating your weights often enough and too small you aren't maximizing benefits of good matrix libraries. Double check that he labelIndex is at 10,000. But the issue is because of the RecordReaderDataSetIterator api you are using. It sets batch size to 10. So when you make the call it will only have 10 examples in the record so it can't split with 80% of 953.

@AldoMarzullo for RecordReaderDataSetIterator you should pass in batchSize vs the total number of pixels in the image. Also not clear how you are using the labels iterator load. You want to append labels in ImageRecordReader based on name in files or pass in a list of labels.
